### ISSUE TITLE
AMDGPUMCInstLower::lower - Pseudo instruction doesn't have a target-specific version: 146

### ISSUE THREAD

**Issue by micah** [mbevand/silentarmy#41]
Thanks for making a linux GPU miner!

When I try to run the sa-solver, I get this output:

```
./sa-solver --nonces 10000 
Solving default all-zero 140-byte header
Building program
OpenCL build failed (-11). Build log follows:
AMDGPUMCInstLower::lower - Pseudo instruction doesn't have a target-specific version: 146
```

I have the amdgpu driver installed, and clinfo shows it:
```
Number of platforms:				 1
  Platform Profile:				 FULL_PROFILE
  Platform Version:				 OpenCL 1.1 Mesa 12.0.3
  Platform Name:				 Clover
  Platform Vendor:				 Mesa
  Platform Extensions:				 cl_khr_icd


  Platform Name:				 Clover
Number of devices:				 1
  Device Type:					 CL_DEVICE_TYPE_GPU
  Vendor ID:					 1002h
  Max compute units:				 36
  Max work items dimensions:			 3
    Max work items[0]:				 256
    Max work items[1]:				 256
    Max work items[2]:				 256
  Max work group size:				 256
  Preferred vector width char:			 16
  Preferred vector width short:			 8
  Preferred vector width int:			 4
  Preferred vector width long:			 2
  Preferred vector width float:			 4
  Preferred vector width double:		 2
  Native vector width char:			 16
  Native vector width short:			 8
  Native vector width int:			 4
  Native vector width long:			 2
  Native vector width float:			 4
  Native vector width double:			 2
  Max clock frequency:				 0Mhz
  Address bits:					 32
  Max memory allocation:			 268435456
  Image support:				 No
  Max size of kernel argument:			 1024
  Alignment (bits) of base address:		 1024
  Minimum alignment (bytes) for any datatype:	 128
  Single precision floating point capability
    Denorms:					 No
    Quiet NaNs:					 Yes
    Round to nearest even:			 Yes
    Round to zero:				 No
    Round to +ve and infinity:			 No
    IEEE754-2008 fused multiply-add:		 No
  Cache type:					 None
  Cache line size:				 0
  Cache size:					 0
  Global memory size:				 1073741824
  Constant buffer size:				 268435456
  Max number of constant args:			 16
  Local memory type:				 Scratchpad
  Local memory size:				 32768
  Kernel Preferred work group size multiple:	 64
  Error correction support:			 0
  Unified memory for Host and Device:		 1
  Profiling timer resolution:			 0
  Device endianess:				 Little
  Available:					 Yes
  Compiler available:				 Yes
  Execution capabilities:				 
    Execute OpenCL kernels:			 Yes
    Execute native function:			 No
  Queue on Host properties:				 
    Out-of-Order:				 No
    Profiling :					 Yes
  Platform ID:					 0x7f640a8d5b00
  Name:						 AMD POLARIS10 (DRM 3.2.0 / 4.7.0-1-amd64, LLVM 3.8.1)
  Vendor:					 AMD
  Device OpenCL C version:			 OpenCL C 1.1 
  Driver version:				 12.0.3
  Profile:					 FULL_PROFILE
  Version:					 OpenCL 1.1 Mesa 12.0.3
  Extensions:					 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_fp64
```

The module is loaded:
```
amdkfd                126976  1
amdgpu               1150976  6
```

```
glxinfo |grep OpenGL
OpenGL vendor string: Advanced Micro Devices, Inc.
OpenGL renderer string: AMD Radeon (TM) RX 480 Graphics
OpenGL core profile version string: 4.5.13453 Core Profile Context
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 4.5.13453 Compatibility Profile Context
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile
OpenGL extensions:
OpenGL ES profile version string: 4.5.13453 Compatibility Profile Context
OpenGL ES profile shading language version string: 4.50
OpenGL ES profile extensions:
```

Any idea what could be causing the problem? 

I'm trying to do this on debian.

**Comment by micah** [comment]
It seems like this error comes from https://smlnj-gitlab.cs.uchicago.edu/manticore/llvm/blob/c75c50f45b3d6d1d61ce6b411d12cedaadd71d5b/lib/Target/AMDGPU/AMDGPUMCInstLower.cpp#L49

**Comment by mbevand** [comment]
It seems you are running the "Clover" OpenCL driver, instead of the AMD OpenCL driver. Can you report the output of:

head -99 /etc/OpenCL/vendors/*
ldd sa-solver
find /usr /opt -name "libOpenCL.so*" 2>/dev/null

**Comment by micah** [comment]
> It seems you are running the "Clover" OpenCL driver, instead of the AMD OpenCL driver. Can you report the output of:
>
> head -99 /etc/OpenCL/vendors/*
> ldd sa-solver
> find /usr /opt -name "libOpenCL.so*" 2>/dev/null

```
silentarmy$ head -99 /etc/OpenCL/vendors/*
==> /etc/OpenCL/vendors/amdocl32.icd <==
libamdocl32.so

==> /etc/OpenCL/vendors/amdocl64.icd <==
libamdocl64.so

==> /etc/OpenCL/vendors/mesa.icd <==
libMesaOpenCL.so.1
micah@muck:silentarmy$ ldd sa-solver 
	linux-vdso.so.1 (0x00007ffcdd5ef000)
	libOpenCL.so.1 => /opt/amdgpu-pro/lib/x86_64-linux-gnu/libOpenCL.so.1 (0x00007f8a46909000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8a4656b000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f8a46363000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8a4605f000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8a45e5b000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8a45c3c000)
	/lib64/ld-linux-x86-64.so.2 (0x00005650e67c4000)
micah@muck:silentarmy$ find /usr /opt -name "libOpenCL.so*" 2>/dev/null
/usr/lib/x86_64-linux-gnu/libOpenCL.so.1
/usr/lib/x86_64-linux-gnu/libOpenCL.so
/opt/AMDAPPSDK-3.0/lib/x86/libOpenCL.so.1
/opt/AMDAPPSDK-3.0/lib/x86/libOpenCL.so
/opt/AMDAPPSDK-3.0/lib/x86_64/sdk/libOpenCL.so.1
/opt/AMDAPPSDK-3.0/lib/x86_64/sdk/libOpenCL.so
/opt/AMDAPPSDK-3.0/lib/x86_64/libOpenCL.so
/opt/amdgpu-pro/lib/i386-linux-gnu/libOpenCL.so.1
/opt/amdgpu-pro/lib/i386-linux-gnu/libOpenCL.so
/opt/amdgpu-pro/lib/x86_64-linux-gnu/libOpenCL.so.1
/opt/amdgpu-pro/lib/x86_64-linux-gnu/libOpenCL.so
silentarmy1$ 
```

**Comment by mbevand** [comment]
mv /etc/OpenCL/vendors/mesa.icd /etc/OpenCL/vendors/mesa.icd.disabled

...then try silentarmy again.

**Comment by micah** [comment]
mbevand <notifications@github.com> writes:

> mv /etc/OpenCL/vendors/mesa.icd /etc/OpenCL/vendors/mesa.icd.disabled
>
> ...then try silentarmy again.

Did that, now I get:

```
silentarmy$ ./sa-solver --nonces 10000 
Solving default all-zero 140-byte header
Cannot get OpenCL platforms (-1001)
```

if I put it back, I get the Pseudo instruction error again

**Comment by mbevand** [comment]
It seems you have a broken install of amdgpu-pro drivers.

Can I ask what steps you took to install it on your Debian system? And is it Jessie?

**Comment by micah** [comment]
mbevand <notifications@github.com> writes:

> It seems you have a broken install of amdgpu-pro drivers.
>
> Can I ask what steps you took to install it on your Debian system? And is it Jessie?

I was afraid of that.

Its a debian sid system. The only linux drivers I could find where for
ubuntu. The install script just tries to detect what release you are,
based on what is in /etc/os-release, only looking for 'ubuntu' or
'steamos' and then installs the packages. So I just made it also check
for for debian.

I'd love to know what is missing or messed up, so it can be corrected!

**Comment by micah** [comment]
mbevand <notifications@github.com> writes:

> It seems you have a broken install of amdgpu-pro drivers.
>
> Can I ask what steps you took to install it on your Debian system? And is it Jessie?

I just solved it... I found that amdgpu-pro_16.40-348864_amd64.deb and
opencl-amdgpu-pro-icd:amd64 were not installed. Once I installed those
and then moved /etc/OpenCL/vendors/mesa.icd out of the way, and now
silentarmy is working (!).

That mesa.icd file comes from mesa-opencl-icd:amd64, so I just removed
that package.
